app/ts/components/TokenAdd.tsx

```diff
@@ -0,0 +1,242 @@
+	)
+}
+
+const QueryAddressField = ({ result }: { result: Signal<Result<EthereumAddress> | undefined> }) => {
+	const query = useSignal('')
+	const isPristine = useSignal<true | undefined>(true)
+	const { ref, signal: inputRef } = useSignalRef<HTMLInputElement | null>(null)
+
+	const parsedAddress = useComputed(() => EthereumAddress.safeParse(query.value))
+
+	const normalizeAndUpdateValue = (newValue: string) => {
+		batch(() => {
+			isPristine.value = undefined
+			query.value = newValue.trim()
+		})
+	}
+
+	const clearValue = () => {
+		if (inputRef.value) {
+			inputRef.value.value = ''
+			const inputEvent = new InputEvent('input')
+			inputRef.value.dispatchEvent(inputEvent)
+			inputRef.value.focus()
+		}
+	}
+
+	const validationMessage = useComputed(() => {
+		if (parsedAddress.value.success) return undefined
+		return 'Invalid ERC20 contract address.'
+	})
+
+	const validateField = () => {
+		if (inputRef.value === null) return
+		if (validationMessage.value === undefined) {
+			inputRef.value.setCustomValidity('')
+			return
+		}
+
+		inputRef.value.setCustomValidity(validationMessage.value)
+		inputRef.value.reportValidity()
+	}
+
+	useSignalEffect(() => {
+		result.value = parsedAddress.value
+	})
+	useSignalEffect(validateField)
+
+	return (
+		<fieldset data-pristine={isPristine.value} class='px-4 py-3 relative grid gap-2 grid-cols-1 grid-flow-col-dense items-center border border-white/50 focus-within:border-white disabled:bg-white/10 disabled:border-white/30 modified:enabled:invalid:border-red-400 group'>
+			<label class='absolute top-2 left-4 text-sm text-white/50 capitalize'>contract address</label>
+			<input ref={ref} type='text' value={query.value} onInput={e => normalizeAndUpdateValue(e.currentTarget.value)} required placeholder='0x0123...' class='peer outline-none pt-4 bg-transparent text-ellipsis disabled:text-white/30 placeholder:text-white/20 group-modified:enabled:invalid:text-red-400' />
+			<ClearButton onClick={clearValue} />
+		</fieldset>
+	)
+}
+
+const QueryResult = ({ result }: { result: Signal<Result<EthereumAddress> | undefined> }) => {
+	const { notify } = useNotice()
+	const { value: query, waitFor, reset } = useAsyncState<ERC20Token>()
+	const { browserProvider, network } = useWallet()
+
+	const getTokenMetadata = () => {
+		if (!result.value?.success) {
+			reset()
+			return
+		}
+
+		if (!browserProvider) {
+			notify({ message: 'No compatible web3 wallet detected.', title: 'Failed to connect' })
+			return
+		}
+
+		if (network.value.state !== 'resolved') {
+			notify({ message: 'Your wallet may not connected to the chain.', title: 'Network unavailable' })
+			return
+		}
+
+		const tokenAddress = result.value.value
+		const activeChainId = network.value.value.chainId
+
+		waitFor(async () => {
+			const contract = new Contract(tokenAddress, ERC20ABI, browserProvider)
+			const name = await contract.name()
+			const symbol = await contract.symbol()
+			const decimals = await contract.decimals()
```

Comment:
```suggestion
			const namePromise = contract.name()
			const symbolPromise = contract.symbol()
			const decimalsPromise = contract.decimals()
			const name = await namePromise
			const symbol = await symbolPromise
			const decimals = await decimalsPromise
```
this would make the calls in paraller